File: src/test/java/com/saucelabs/tests/TestLogin.java
package com.saucelabs.tests;

import com.saucelabs.config.ConfigReader;
import com.saucelabs.pages.LoginPage;
import com.saucelabs.pages.ProductsPage;
import org.testng.Assert;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

/**
 * TestLogin - Test cases for the Login page.
 * Covers: valid login, invalid credentials, locked user, empty fields, UI elements.
 */
public class TestLogin extends BaseTest {

    private static final String BASE_URL     = ConfigReader.getBaseUrl();
    private static final String VALID_USER   = ConfigReader.getValidUsername();
    private static final String VALID_PASS   = ConfigReader.getValidPassword();
    private static final String LOCKED_USER  = ConfigReader.getLockedUsername();

    @BeforeMethod(alwaysRun = true)
    @Override
    public void setUp(java.lang.reflect.Method method) {
        super.setUp(method); // creates the driver; no login needed for login tests
    }

    @Test(groups = {"login", "smoke"}, description = "TC_LOGIN_001: Verify successful login with valid credentials.")
    public void testValidLogin() {
        LoginPage loginPage = new LoginPage(driver);
        loginPage.navigateTo(BASE_URL);
        loginPage.login(VALID_USER, VALID_PASS);

        ProductsPage productsPage = new ProductsPage(driver);
        Assert.assertTrue(productsPage.isProductsPageDisplayed(),
                "Products page not displayed after valid login");
        Assert.assertTrue(driver.getCurrentUrl().contains("inventory"),
                "URL should contain 'inventory'");
    }

    @Test(groups = {"login", "smoke"}, description = "TC_LOGIN_002: Verify error message with invalid username.")
    public void testInvalidUsername() {
        LoginPage loginPage = new LoginPage(driver);
        loginPage.navigateTo(BASE_URL);
        loginPage.login("invalid_user", VALID_PASS);

        Assert.assertTrue(loginPage.isErrorDisplayed(), "Error message should be displayed");
        Assert.assertTrue(loginPage.getErrorMessage().contains("Username and password do not match"),
                "Unexpected error message");
    }

    @Test(groups = {"login", "smoke"}, description = "TC_LOGIN_003: Verify error message with invalid password.")
    public void testInvalidPassword() {
        LoginPage loginPage = new LoginPage(driver);
        loginPage.navigateTo(BASE_URL);
        loginPage.login(VALID_USER, "wrong_password");

        Assert.assertTrue(loginPage.isErrorDisplayed(), "Error message should be displayed");
        Assert.assertTrue(loginPage.getErrorMessage().contains("Username and password do not match"),
                "Unexpected error message");
    }

    @Test(groups = {"login", "smoke"}, description = "TC_LOGIN_004: Verify locked-out user gets appropriate error.")
    public void testLockedOutUser() {
        LoginPage loginPage = new LoginPage(driver);
        loginPage.navigateTo(BASE_URL);
        loginPage.login(LOCKED_USER, VALID_PASS);

        Assert.assertTrue(loginPage.isErrorDisplayed(), "Error message should be displayed");
        Assert.assertTrue(loginPage.getErrorMessage().toLowerCase().contains("locked out"),
                "Expected 'locked out' in error message: " + loginPage.getErrorMessage());
    }

    @Test(groups = {"login", "smoke"}, description = "TC_LOGIN_005: Verify error when submitting empty credentials.")
    public void testEmptyCredentials() {
        LoginPage loginPage = new LoginPage(driver);
        loginPage.navigateTo(BASE_URL);
        loginPage.clickLogin();

        Assert.assertTrue(loginPage.isErrorDisplayed(), "Error message should be displayed");
        Assert.assertTrue(loginPage.getErrorMessage().contains("Username is required"),
                "Expected 'Username is required' error");
    }

    @Test(groups = {"login", "smoke"}, description = "TC_LOGIN_006: Verify all UI elements are present on login page.")
    public void testLoginPageUiElements() {
        LoginPage loginPage = new LoginPage(driver);
        loginPage.navigateTo(BASE_URL);

        Assert.assertTrue(loginPage.isLoginPageDisplayed(), "Login page not displayed");
        Assert.assertEquals(loginPage.getLogoText(), "Swag Labs", "Logo text mismatch");
        Assert.assertTrue(loginPage.areCredentialsDisplayed(), "Credentials section missing");
        Assert.assertTrue(loginPage.isPasswordHintDisplayed(), "Password hint section missing");
    }
}

File: saucelabs-selenium-project/tests/test_login.py
"""
test_login.py - Test cases for the Login page.
Covers: valid login, invalid credentials, locked user, empty fields, UI elements.
"""
import pytest
from pages.login_page import LoginPage
from pages.products_page import ProductsPage
from utils.config_reader import ConfigReader

BASE_URL = ConfigReader.get_base_url()
VALID_USER = ConfigReader.get_valid_username()
VALID_PASS = ConfigReader.get_valid_password()
LOCKED_USER = ConfigReader.get_locked_username()


@pytest.mark.login
@pytest.mark.smoke
class TestLogin:
    """Test suite for Login functionality."""

    def test_valid_login(self, driver):
        """TC_LOGIN_001: Verify successful login with valid credentials."""
        login_page = LoginPage(driver)
        login_page.navigate_to(BASE_URL)
        login_page.login(VALID_USER, VALID_PASS)

        products_page = ProductsPage(driver)
        assert products_page.is_products_page_displayed(), "Products page not displayed after valid login"
        assert "inventory" in driver.current_url, "URL should contain 'inventory'"

    def test_invalid_username(self, driver):
        """TC_LOGIN_002: Verify error message with invalid username."""
        login_page = LoginPage(driver)
        login_page.navigate_to(BASE_URL)
        login_page.login("invalid_user", VALID_PASS)

        assert login_page.is_error_displayed(), "Error message should be displayed"
        error_msg = login_page.get_error_message()
        assert "Username and password do not match" in error_msg

    def test_invalid_password(self, driver):
        """TC_LOGIN_003: Verify error message with invalid password."""
        login_page = LoginPage(driver)
        login_page.navigate_to(BASE_URL)
        login_page.login(VALID_USER, "wrong_password")

        assert login_page.is_error_displayed(), "Error message should be displayed"
        error_msg = login_page.get_error_message()
        assert "Username and password do not match" in error_msg

    def test_locked_out_user(self, driver):
        """TC_LOGIN_004: Verify locked-out user gets appropriate error."""
        login_page = LoginPage(driver)
        login_page.navigate_to(BASE_URL)
        login_page.login(LOCKED_USER, VALID_PASS)

        assert login_page.is_error_displayed(), "Error message should be displayed"
        error_msg = login_page.get_error_message()
        assert "locked out" in error_msg.lower(), f"Expected 'locked out' in: {error_msg}"

    def test_empty_credentials(self, driver):
        """TC_LOGIN_005: Verify error when submitting empty credentials."""
        login_page = LoginPage(driver)
        login_page.navigate_to(BASE_URL)
        login_page.click_login()

        assert login_page.is_error_displayed(), "Error message should be displayed"
        error_msg = login_page.get_error_message()
        assert "Username is required" in error_msg

    def test_login_page_ui_elements(self, driver):
        """TC_LOGIN_006: Verify all UI elements are present on login page."""
        login_page = LoginPage(driver)
        login_page.navigate_to(BASE_URL)

        assert login_page.is_login_page_displayed(), "Login page not displayed"
        assert login_page.get_logo_text() == "Swag Labs", "Logo text mismatch"
        assert login_page.are_credentials_displayed(), "Credentials section missing"
        assert login_page.is_password_hint_displayed(), "Password hint section missing"
